src/app/codegen-interaction-model/CodegenDataModel.h

```diff
@@ -0,0 +1,84 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/interaction-model/Model.h>
+
+#include <app/util/af-types.h>
+
+namespace chip {
+namespace app {
+
+class CodegenDataModel : public chip::app::InteractionModel::Model
+{
+public:
+    /// Generic model implementations
+    CHIP_ERROR Shutdown() override { return CHIP_NO_ERROR; }
+
+    CHIP_ERROR ReadAttribute(const InteractionModel::ReadAttributeRequest & request, InteractionModel::ReadState & state,
+                             AttributeValueEncoder & encoder) override;
+    CHIP_ERROR WriteAttribute(const InteractionModel::WriteAttributeRequest & request, AttributeValueDecoder & decoder) override;
+    CHIP_ERROR Invoke(const InteractionModel::InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
+                      InteractionModel::InvokeReply & reply) override;
+
+    /// attribute tree iteration
+    EndpointId FirstEndpoint() override;
+    EndpointId NextEndpoint(EndpointId before) override;
```

Comment:
What's the reasoning behind this style of iteration API, compared to following the STL iterator paradigm?

Reply:
I could not figure out a good API for this. Ideally I would want a `EndpointIterator` that iterates over things (and for clusters and attributes the same). However I also expect the actual implementation to need to be hidden and if I make the iterator have virtual bits I start going into dynamic memory allocation (or at least pooled one with ReferenceCounted). I also need to work a lot harder on invalidation probably (so I would go into some weak-pointer-style support). After thinking a bit about that I thought I go too much down the  rabbit hole and I left it at this style.

Eventually this could be possible, however for now this seemed good enough to move forward for now.